File: XPile 3-17/Test/test_types.py
from dataclasses import dataclass, field

@dataclass
class Point:
    x: int = 0
    y: int = 0

@dataclass
class Person:
    name: str = ""
    age: int = 0
    location: Point = field(default_factory=Point)

def print_point(p):
    print(p.x)
    print(p.y)

def make_point(x, y):
    result = Point()
    result.x = x
    result.y = y
    return result

def main():
    p = Point()
    p.x = 10
    p.y = 20
    print(p.x)
    print(p.y)
    bob = Person()
    bob.name = 'Bob'
    bob.age = 30
    bob.location.x = 100
    bob.location.y = 200
    print(bob.name)
    print(bob.age)
    print(bob.location.x)
    print(bob.location.y)
    print_point(p)
    q = Point()
    q = make_point(5, 15)
    print(q.x)
    print(q.y)


if __name__ == "__main__":
    main()

File: XPile 3-17/Test/test_types.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

typedef struct {
    int x;
    int y;
} Point;

typedef struct {
    char* name;
    int age;
    Point location;
} Person;

void print_point(Point p);
Point make_point(int x, int y);
void main(void);

void print_point(Point p)
{
    printf("%d\n", p.x);
    printf("%d\n", p.y);
}

Point make_point(int x, int y)
{
    Point result;
    result.x = x;
    result.y = y;
    return result;
}

void main(void)
{
    Point p;
    p.x = 10;
    p.y = 20;
    printf("%d\n", p.x);
    printf("%d\n", p.y);
    Person bob;
    bob.name = "Bob";
    bob.age = 30;
    bob.location.x = 100;
    bob.location.y = 200;
    printf("%s\n", bob.name);
    printf("%d\n", bob.age);
    printf("%d\n", bob.location.x);
    printf("%d\n", bob.location.y);
    print_point(p);
    Point q;
    q = make_point(5, 15);
    printf("%d\n", q.x);
    printf("%d\n", q.y);
}

